hivemind/averaging/control.py

```diff
@@ -0,0 +1,136 @@
+import struct
+from enum import Enum
+from typing import Optional
+
+import numpy as np
+import torch
+
+from hivemind.utils import DHTExpiration, MPFuture, get_dht_time, get_logger
+
+logger = get_logger(__name__)
+
+
+class AveragingStage(Enum):
+    IDLE = 0  # still initializing
+    LOOKING_FOR_GROUP = 1  # running decentralized matchmaking, can't run allreduce yet
+    AWAITING_TRIGGER = 2  # waiting for user to set the trigger that allows running allreduce
+    RUNNING_ALLREDUCE = 3  # exchanging tensors with groupmates
+    FINISHED = 4  # either done or failed with exception
+
+
+class StepControl(MPFuture):
+    """
+    An auxiliary data structure that allows user to control stages and track progress in a single averaging step
+
+    :param scheduled_time: estimated time when averaging should begin. Will be used for scheduling
+    :param deadline: if averaging is still in progress at this time, it should be stopped due to TimeoutError
+    :param allow_retries: if True, allow running matchmaking and all-reduce again if previous attempt fails
+    :param weight: averaging weight, can be changed afterwards
+    :param gather_binary: optionally send this data to all peers in the next group and gather it from groupmates
+    :returns: an assembled group if successful, None if failed; does NOT perform the actual averaging
+    """
+
+    def __init__(
+        self, scheduled_time: DHTExpiration, deadline: float, allow_retries: bool, weight: float, gather_binary: bytes
```

Comment:
```suggestion
        self, scheduled_time: DHTExpiration, deadline: float, allow_retries: bool, weight: float, gather_binary: Optional[bytes]
```

Reply:
this argument is not optional: even if user does not supply data, it will be serialized as bytes alongside (bandwidth, averaging_mode) 

https://github.com/learning-at-home/hivemind/blob/pre_scheduling_again/hivemind/averaging/averager.py#L364-L365

Reply:
Got it, the reason why I suggested it is that it's described as optional in the docstring

Reply:
fixd